### DISCUSSION TITLE
Step file Joint Data

### DISCUSSION THREAD

**Discussion by shrekuntu** [gumyr/build123d#769] [General]
Here is an example of a simple cylinder that I need to attach  flanges on both ends symmetrically (mirrored).
Flanges need to be imported from step file.

First I make flanges with Joint data and export to step:

```
from build123d import *

with BuildPart() as c:
    with BuildSketch(Plane.YZ) as sk:
        Circle(8 * IN)
    extrude(amount=1 * IN, taper=30)
    
    RigidJoint(
        label="flange",
        joint_location=Location((0,0,0), (0, 0, 0))
    )


export_step(to_export=c.part, file_path="c.step")

```

then I make a cyclinder with two RigidJoint on each end to connect flanges imported from step file

```
from build123d import *
r = 5 * IN
h = 20 * IN

with BuildPart(Plane.YZ) as p:
    Cylinder(radius=r, height=h)

    RigidJoint(
        label="left",
        joint_location=Location((-h / 2 - 1 * IN,0,0), (0, 0, 180))
    )
    RigidJoint(
        label="right",
        joint_location=Location((h / 2+ 1 * IN,0,0), (0, 0, 0))
    )
 
c = import_step("c.step")

p.part.joints['left'].connect_to(c.joints['flange'])
p.part.joints['right'].connect_to(c2.joints['flange'])
```

but I get an error: KeyError: 'flange'. I can see that nowhere in the step file there is a work flange.
What is the best way to do this?

**Comment by gumyr** [comment]
The STEP file format doesn't support `Joints` that is a build123d concept exclusively. Is there a reason why you are trying to store your parts in a STEP file?

**Reply by shrekuntu** [reply]
I showed a simple example, but my real parts are pulley, hubs and bushings. I premade hubs and bushings from database records and when I'm building a pulley I was thinking the best way is to bring them in from a step file and constrain accordingly.
I am open for any suggestion, it doesn't have to be  step file.
I was studying your pipe-flanges code where you use joints, but how do flanges from bd-warehouse know their joints data? Are they always generated from scratch?

**Reply by gumyr** [reply]
Yes, the flanges are generated on the fly, they aren't stored anywhere (there are so many possibilities that's impractical).

**Reply by shrekuntu** [reply]
I will try it on the fly. Thank you @gumyr for your help.

**Reply by gumyr** [reply]
If you have a library of parts that you'd like to share let me know and it could be added to the build123d docs.

**Reply by shrekuntu** [reply]
Thanks, I will once my code is more polished, right now it's still rough.
